### PR TITLE
Integration tests for local slam.

### PR DESCRIPTION
**Pull request by gaschler**

### CODE REVIEW

File: cartographer/mapping/map_builder_test.cc [hunk 1]
```diff
@@ -31,6 +31,37 @@ namespace cartographer {
 namespace mapping {
 namespace {
 
+const std::string kRangeSensorId = "range";
+const std::string kIMUSensorId = "imu";
+const double kDuration = 2.;
```
Comment (cschuet) [inline cartographer/mapping/map_builder_test.cc]: constexpr

File: cartographer/mapping/map_builder_test.cc [hunk 2]
```diff
@@ -31,6 +31,37 @@ namespace cartographer {
...
+const double kTimeStep = 0.1;
+
+struct FakeRangeMeasurement {
```
Comment (cschuet) [inline cartographer/mapping/map_builder_test.cc]: My CL #736 actually introduces a data structure for this.

File: cartographer/mapping/map_builder_test.cc [hunk 3]
```diff
@@ -31,6 +31,37 @@ namespace cartographer {
...
+  common::Time time;
+  sensor::TimedPointCloud ranges;
+};
+
+std::vector<FakeRangeMeasurement> GenerateFakeRangeMeasurements() {
+  std::vector<FakeRangeMeasurement> measurements;
+  sensor::TimedPointCloud point_cloud;
+  for (double angle = 0.; angle < 1. * M_PI; angle += 0.01) {
+    const double kRadius = 5;
```
Comment (cschuet) [inline cartographer/mapping/map_builder_test.cc]: constexpr

File: cartographer/mapping/map_builder_test.cc [hunk 4]
```diff
@@ -31,6 +31,37 @@ namespace cartographer {
...
+    point_cloud.emplace_back(kRadius * std::cos(angle),
+                             kRadius * std::sin(angle), 0., 0.);
+  }
+  for (double elapsed_time = 0.; elapsed_time < kDuration;
+       elapsed_time += kTimeStep) {
+    common::Time time =
+        common::FromUniversal(123) + common::FromSeconds(elapsed_time);
```
Comment (cschuet) [inline cartographer/mapping/map_builder_test.cc]: Is it important for the experiment to start at 123? Why not omit this entirely?

File: cartographer/mapping/map_builder_test.cc [hunk 5]
```diff
@@ -31,6 +31,37 @@ namespace cartographer {
...
+    transform::Rigid3f pose = transform::Rigid3f::Translation(
+        Eigen::Vector3f(elapsed_time * 0.2, elapsed_time * 0.1, 0.));
```
Comment (cschuet) [inline cartographer/mapping/map_builder_test.cc]: Introduce a Rigid{2,3}f as a velocity vector and name accordingly?

File: cartographer/mapping/map_builder_test.cc [hunk 6]
```diff
@@ -48,37 +79,125 @@ class MapBuilderTest : public ::testing::Test {
     const std::string kMapBuilderLua = R"text(
       include "map_builder.lua"
       MAP_BUILDER.use_trajectory_builder_2d = true
+      MAP_BUILDER.pose_graph.optimize_every_n_nodes = 0
```
Comment (cschuet) [inline cartographer/mapping/map_builder_test.cc]: Can you add a comment indicating that this disables global SLAM?

File: cartographer/mapping/map_builder_test.cc [hunk 7]
```diff
@@ -48,37 +79,125 @@ class MapBuilderTest : public ::testing::Test {
...
+                                           Eigen::Vector3f::Zero(),
+                                           measurement.ranges);
+    trajectory_builder->AddImuData(kIMUSensorId, measurement.time,
+                                   Eigen::Vector3d(0., 0., 9.8),
+                                   Eigen::Vector3d::Zero());
+  }
+  map_builder_->FinishTrajectory(trajectory_id);
+  map_builder_->pose_graph()->RunFinalOptimization();
+  EXPECT_GE(local_slam_result_poses_.size(), 5);
+  EXPECT_GE((local_slam_result_poses_.back().translation() -
```
Comment (cschuet) [inline cartographer/mapping/map_builder_test.cc]: Introduce upper and lower bound?

Comment (cschuet) [reply cartographer/mapping/map_builder_test.cc]: EXPECT_NEAR

File: cartographer/mapping/map_builder_test.cc [hunk 8]
```diff
@@ -48,37 +79,125 @@ class MapBuilderTest : public ::testing::Test {
...
+      map_builder_->GetTrajectoryBuilder(trajectory_id);
+  for (const auto& measurement : GenerateFakeRangeMeasurements()) {
+    trajectory_builder->AddRangefinderData(kRangeSensorId, measurement.time,
+                                           Eigen::Vector3f::Zero(),
+                                           measurement.ranges);
+  }
+  map_builder_->FinishTrajectory(trajectory_id);
+  map_builder_->pose_graph()->RunFinalOptimization();
+  EXPECT_GE(local_slam_result_poses_.size(), 5);
+  EXPECT_GE((local_slam_result_poses_.back().translation() -
```
Comment (cschuet) [inline cartographer/mapping/map_builder_test.cc]: EXPECT_NEAR

File: cartographer/mapping/map_builder_test.cc (hunk 4 above)
Comment (gaschler) [reply cartographer/mapping/map_builder_test.cc]: It is necessary to give an absolute time (FromSeconds only outputs a duration), and I prefer to choose something different from zero.

**Comment by gaschler** [comment]
Done with the corrections, only waiting to use TimedPointCloudData from PR #736.

**Pull request merged**